### PR TITLE
Non blocking/7 select subscriptions

### PR DESCRIPTION
**Pull request by wangmingliang-ms**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java [hunk 1]
```diff
@@ -108,11 +76,40 @@ public void update(AnActionEvent e) {
...
+
+        return loadSubscriptions(subscriptionManager, project)
+            .switchMap((list) -> selectSubscriptions(project, subscriptionManager))
+            .toSingle()
+            .doOnSuccess(subscriptionManager::setSubscriptionDetails);
+    }
+
+    private static Observable<List<SubscriptionDetail>> selectSubscriptions(final Project project, final SubscriptionManager subscriptionManager) {
+        return AzureTaskManager.getInstance().runLater(new AzureTask<>(() -> {
+            //System.out.println("onShowSubscriptions: calling getSubscriptionDetails()");
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java:98]: remove this line

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java [hunk 2]
```diff
@@ -59,43 +60,10 @@ public SelectSubscriptionsAction() {
     @AzureOperation(value = "select one/more subscriptions to work on", type = AzureOperation.Type.ACTION)
     public boolean onActionPerformed(@NotNull AnActionEvent e, @Nullable Operation operation) {
         Project project = DataKeys.PROJECT.getData(e.getDataContext());
-        onShowSubscriptions(project);
+        selectSubscriptions(project).subscribe();
         return true;
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java:64]: can we put subscribe() method into selectSubscriptions(...) ?
I think it‘s elusive for developers to call it.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java [hunk 3]
```diff
@@ -108,11 +76,40 @@ public void update(AnActionEvent e) {
         }
     }
 
-    public static void updateSubscriptionWithProgressDialog(final SubscriptionManager subscriptionManager, Project project) {
-        AzureTaskManager.getInstance().runInModal(new AzureTask(project, "Loading Subscriptions...", true, () -> {
+    public static Single<List<SubscriptionDetail>> selectSubscriptions(Project project) {
+        //Project project = ProjectManager.getInstance().getDefaultProject();();
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java:80]: remove the unnecessary line ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java [hunk 4]
```diff
@@ -108,11 +76,40 @@ public void update(AnActionEvent e) {
...
+            .toSingle()
+            .doOnSuccess(subscriptionManager::setSubscriptionDetails);
+    }
+
+    private static Observable<List<SubscriptionDetail>> selectSubscriptions(final Project project, final SubscriptionManager subscriptionManager) {
+        return AzureTaskManager.getInstance().runLater(new AzureTask<>(() -> {
+            //System.out.println("onShowSubscriptions: calling getSubscriptionDetails()");
+            SubscriptionsDialog d = SubscriptionsDialog.go(subscriptionManager.getSubscriptionDetails(), project);
+            List<SubscriptionDetail> subscriptionDetailsUpdated;
+            return d != null ? d.getSubscriptionDetails() : null;
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java:101]: Objects.nonNull

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java [hunk 5]
```diff
@@ -108,11 +76,40 @@ public void update(AnActionEvent e) {
...
+        }));
+    }
+
+    @AzureOperation(value = "load all available subscriptions from Azure", type = AzureOperation.Type.SERVICE)
+    public static Observable<List<SubscriptionDetail>> loadSubscriptions(final SubscriptionManager subscriptionManager, Project project) {
+        return AzureTaskManager.getInstance().runInModal(new AzureTask<>(project, "Loading Subscriptions...", true, () -> {
             ProgressManager.getInstance().getProgressIndicator().setIndeterminate(true);
             //System.out.println("updateSubscriptionWithProgressDialog: calling getSubscriptionDetails()");
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java:109]: remove the line as well ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java (hunk 2 above)
Comment (wangmingliang-ms) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java:64]: no, some other callings need to do something in `subscribe`, e.g.
image

Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/actions/SelectSubscriptionsAction.java:65]: how about adding anoter selectSubscriptions(myProject, subActions) ?

**Pull request merged**